storage: bugfixes

brauner opened the pull request.

stgraber commented:
jenkins: test this please

stgraber commented on lxd/container.go:
```diff
@@ -256,6 +257,15 @@ func containerValidDevices(devices types.Devices, profile bool, expanded bool) e
 		} else if m["type"] == "disk" {
+			if !expanded && !shared.StringInSlice(m["path"], diskDevicePaths) {
+				diskDevicePaths = append(diskDevicePaths, m["path"])
+			} else if !expanded && len(diskDevicePaths) > 0 {
```
This reads a bit funny, why is len(diskDevicePaths) needed at all here?

brauner replied:
Removed.

brauner commented on lxd/container_put.go:
```diff
@@ -39,6 +40,79 @@ func containerPut(d *Daemon, r *http.Request) Response {
+	for _, v := range configRaw.Devices {
+		// Detect whether the new profile has a root device.
+		if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && v["pool"] != "" {
+			newLocalRootDevice = v
```
Needs a break statement.

brauner commented on lxd/container_put.go:
```diff
@@ -39,6 +40,79 @@ func containerPut(d *Daemon, r *http.Request) Response {
+	var oldLocalRootDevice types.Device
+	for _, v := range c.LocalDevices() {
+		// Detect whether the old profile has a root device.
+		if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" && v["pool"] != "" {
```
Needs a break statement.

brauner commented:
Somehow jenkins is either not run on this branch or it is not shown.

stgraber commented:
@brauner Github says there's a conflict in test/main.sh somehow. Rebase on master?

brauner commented:
One sec, my rebase somehow got messed up.

brauner commented:
Weird.

stgraber commented on lxd/container_lxc.go:
```diff
@@ -3224,54 +3226,34 @@ func (c *containerLXC) Update(args containerArgs, userRequested bool) error {
+	// Check whether the old local devices have a root disk device.
+	oldRootDiskDevice := containerGetRootDiskDevice(oldExpandedDevices)
+	if oldRootDiskDevice == nil || oldRootDiskDevice["pool"] == "" {
+		return fmt.Errorf("No root disk device found.")
```
Doesn't that effectively prevent you from ever fixing the problem, should a container be in that state already?

brauner replied:
The new branch will allow for "fixing" broken containers.

stgraber commented on lxd/container_put.go:
```diff
@@ -39,6 +39,60 @@ func containerPut(d *Daemon, r *http.Request) Response {
+		} else if oldLocalRootDevice["pool"] == "" && newLocalRootDevice["pool"] != profileRootDiskDevices[0] {
+			return BadRequest(fmt.Errorf("The update would change the containers' storage pool."))
+		}
+	}
```
I don't get why we're adding all this complexity here instead of doing the checks in Update()

Doesn't doing all this here also means that Patch() isn't covered?

stgraber replied:
Would also be a bunch simpler from inside Update() since we have the old and new expanded devices readily available, so no need to load all the profiles ourselves.

brauner replied:
The checks have now been moved into Update(). They indeed make a lot more sense over there.

stgraber commented on lxd/containers_post.go:
```diff
@@ -432,6 +432,26 @@ func containersPost(d *Daemon, r *http.Request) Response {
+
+	if len(profileRootDiskDevices) > 1 && newLocalRootDevice == nil || (len(profileRootDiskDevices) > 1 && newLocalRootDevice["pool"] == "") {
+		return BadRequest(fmt.Errorf("Profiles specify contradicting storage pools the container belongs to."))
+	}
```
Wouldn't this check be better suited for containerCreateInternal which is where we already run all the other config checks?

brauner replied:
Right.

stgraber commented on lxd/profiles_utils.go:
```diff
@@ -0,0 +1,147 @@
+	if len(containers) > 0 && oldProfileRootDiskDevice != nil && oldProfileRootDiskDevice["pool"] != "" {
+		if newProfileRootDiskDevice == nil || newProfileRootDiskDevice["pool"] == "" || (oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+			// Check for any containers that have a local root disk device.
+			for i := 0; i < len(containers); i++ {
```
for _, container := range containers {

brauner replied:
I'm using the index to delete containers from the slice.

stgraber commented on lxd/profiles_utils.go:
```diff
@@ -0,0 +1,147 @@
+		if newProfileRootDiskDevice == nil || newProfileRootDiskDevice["pool"] == "" || (oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+			// Check for any containers that have a local root disk device.
+			for i := 0; i < len(containers); i++ {
+				for _, v := range containers[i].LocalDevices() {
```
for _, v := range container.LocalDevices()

brauner replied:
I'm using the index to delete containers from the slice.

stgraber commented on lxd/profiles_utils.go:
```diff
@@ -0,0 +1,147 @@
+				return BadRequest(fmt.Errorf("At least one container relies on this profiles' root disk device."))
+			}
+		}
+	}
```
This whole chunk of code looks much more complicated that it should be.

What exactly are you trying to do here?

stgraber replied:
So the goal is to know whether any container relies on this profile's root device and if that's the case, prevent changes to it.

```
// Check for containers using the device
for _, container := range containers {
    // Check if the device is locally overridden 
    _, ok := container.Devices[device_name]
    if ok {
        continue
    }

    // Check what profile the device comes from
    profiles = container.Profiles()
    for i := len(profiles)-1; i >= 0; i-- {
        _, profile, err := dbProfileGet(d.db, profiles[i])
        if err != nil {
            return err
        }

        // Found the profile
        _, ok := profile.Devices[device_name]
        if ok {
            if profiles[i] == profile_name {
                return fmt.Errorf("At least one container relies on this profile's root disk device.")
            } else {
                break
            }
        }
    }
}
```

That should achieve the same goal and seems a bit more readable to me.

brauner replied:
You can't rely on the root disk device to have the same name in the local devices map and in the profile devices map.

brauner replied:
But I might be wrong. In any case, I'm pushing a new version and we can change that too if my argument is wrong.

brauner replied:
I simplified the code a little too.

stgraber replied:
What we need to do during validation is:
 - Check that the expanded configuration doesn't contain two entries for any given "path" property.
 - With that validated, then only a single device can be the container's root device
 - As profiles override each other based on device names and container's overrides work the same way (masks the device based on its name), all we need to check is what's the source of the container's root disk entry. That's what my proposed code does by going through the sorted list of profiles in reverse order and checking for the first which defines that device's name. If the profile it finds is the one being modified, then that means that changes to the profile would affect the container.

What am I missing?

brauner commented:
@stgraber, so I agree with moving the container checks to update but I fear some of the complexity in `profiles_update.go` has to stay.

brauner commented:
Github is doing weird stuff. It's reordering my comments and also introduces random merge conflicts.

stgraber commented on lxd/container_lxc.go:
```diff
@@ -1440,32 +1441,42 @@ func (c *containerLXC) expandDevices() error {
-		c.storagePool = profileStoragePool
 	}
 
 	c.expandedDevices = devices
```
So I really don't think we want expandConfig or expandDevices to do ANY config/devices validation.
Those functions are called constantly from all over the place and should be very minimal and very quick.

The only times where the container's configuration should be validate is when the container is created or when the container is changed. If we do it at any other time, we risk re-introducing those kind of problems where even trivial things like "lxc config show" now fail, making it impossible for the user to recover unless they do DB surgery.

brauner replied:
ok

brauner replied:
removed

stgraber commented:
Copy/pasting my comment about profile data validation here since Github is hiding it...

What we need to do during validation is:
 - Check that the expanded configuration doesn't contain two entries for any given "path" property.
 - With that validated, then only a single device can be the container's root device
 - As profiles override each other based on device names and container's overrides work the same way (masks the device based on its name), all we need to check is what's the source of the container's root disk entry. That's what my proposed code does by going through the sorted list of profiles in reverse order and checking for the first which defines that device's name. If the profile it finds is the one being modified, then that means that changes to the profile would affect the container.

What am I missing?

brauner commented:
Sorry, what do you mean override each other based on device names?

So this is the first time I have to deal with profiles in any deeper sense. What do you mean override each other based on device names? What I thought was: The expanded configuration consists of the local devices and all profile devices. So,
- if I have a disk device called "root" in the local devices of the container and a disk device called "root" in the profiles devices the expanded container devices only contain one device "root" as they override each other even though they might be of different types?
- If I have disk device called "root" in the local devices of the container and a disk device called "root2" in the profile devices, that both provide identical information (e.g. path=/ type=disk pool=default) or different information wrt to the pool (e.g. path=/ type=disk but "root" has pool=default and "root2" has pool=pool1) then both devices will show up in the expanded devices, right?

stgraber commented:
@brauner yup, that's correct.

Which is why the first point in what I listed is to check expandedDevices to make sure we don't have two devices with the same "path" property.

Once that's out of the way, we're dealing with a single "device" entry. So get the name of that device entry and find where it comes from. If it comes from the profile we're modifying, then we can't allow the profile to be modified. If it comes from some other profile that's overriding our current profile, then we're good to modify the profile as it wouldn't impact it.

brauner commented:
@stgraber, right. Cool. One more, if we're going to remove the detection code in `expandDevices()` (which I'm very much in favour, then I'd like to give this struct:

```
type containerArgs struct {
	// Don't set manually
	Id int

	Architecture int
	BaseImage    string
	Config       map[string]string
	CreationDate time.Time
	LastUsedDate time.Time
	Ctype        containerType
	Devices      types.Devices
	Ephemeral    bool
	Name         string
	Profiles     []string
	Stateful     bool
}
```

and additional argument `StoragePool string` which is mainly used during container creation. This is what I had implemented before we switched to the `expandDevices()` stuff. I am strongly favoring this additional argument approach as it lines up nicely with storage pool validation we now do in `containerCreateInternal()`. So we'd have there something like:

```
	// Check that there are no contradicting root disk devices.
	var profileRootDiskDevices []string
	for _, pName := range args.Profiles {
		_, p, err := dbProfileGet(d.db, pName)
		if err != nil {
			return nil, fmt.Errorf("Could not load profile '%s'.", pName)
		}

		k, v := containerGetRootDiskDevice(p.Devices)
		if k != "" && !shared.StringInSlice(v["pool"], profileRootDiskDevices) {
			profileRootDiskDevices = append(profileRootDiskDevices, v["pool"])
		}
	}

	_, newLocalRootDiskDevice := containerGetRootDiskDevice(args.Devices)
        args.StoragePool = newLocalRootDiskDevice["pool"]
	if newLocalRootDiskDevice["pool"] == "" {
		if len(profileRootDiskDevices) == 0 {
			return nil, fmt.Errorf("Container relies on profile's root disk device but none was found")
		} else if len(profileRootDiskDevices) > 1 {
			return nil, fmt.Errorf("Container relies on profile's root disk device but conflicting devices were found")
		}
                args.StoragePool = profileRootDiskDevices[0]
	}

```

and the pool then get's passed to `containerLXCCreate()` I find this to be much cleaner than what we do now.

stgraber commented:
I'm not sure I understand what we'd get from adding a StoragePool member to the containerArgs struct. That struct member wouldn't align with anything in the API or the database as the API to set a pool directly on a container is to use a device in the Devices list.

brauner commented:
@stgraber, it's mostly convenience. Otherwise we can add a custom function that retrieves the root device from the devices list and call it in `containerLXCCreate()` and friends.

stgraber commented:
@brauner I think I'd prefer a simple function which just does:

```
for _, d := range c.ExpandedDevices() {
    if d["type"] == "disk" && d["path"] == "/" {
        return d["pool"]
    }
}
```

brauner commented:
@stgraber, yeah we already have that. I was thinking:

```
// Initialize the storage pool for this container.
func (c *containerLXC) initStoragePool() error {
	if c.storagePool != "" {
		return nil
	}

	_, rootDiskDevice := containerGetRootDiskDevice(c.localDevices)
	if rootDiskDevice["pool"] == "" {
		_, rootDiskDevice = containerGetRootDiskDevice(c.expandedDevices)
	}
	if rootDiskDevice["pool"] == "" {
		return fmt.Errorf("No storage pool found in the containers devices.")
	}
	c.storagePool = rootDiskDevice["pool"]

	return nil
}
```

brauner commented on lxd/profiles_utils.go:
```diff
@@ -0,0 +1,135 @@
+			// Check if the device is locally overridden
+			localDevices := container.LocalDevices()
+			k, v := containerGetRootDiskDevice(localDevices)
+			if k != "" && v["pool"] != "" {
```
So I think this is what we want here:
- containerValidDevices() checks that there are no entries with the same path.
- now we check for a local device in the containers local devices map.
- if we find local device with an appropriate pool, we go to the next container
- otherwise we do the check you suggested, looking whether there is another profile that provides the same root disk device

I just want to add another test case to the suite, that we cannot assign profiles to a container that specify conflicting root disk devices when it relies on a profile for its root disk device.

stgraber replied:
I think that's right

brauner replied:
Added the test I mentioned

stgraber commented on lxd/container.go:
```diff
@@ -660,6 +679,29 @@ func containerCreateInternal(d *Daemon, args containerArgs) (container, error) {
+
+		k, v := containerGetRootDiskDevice(p.Devices)
+		if k != "" && !shared.StringInSlice(v["pool"], profileRootDiskDevices) {
+			profileRootDiskDevices = append(profileRootDiskDevices, v["pool"])
```
Don't you mean to append "k" to that list instead of the pool name?

stgraber commented on lxd/container.go:
```diff
@@ -660,6 +679,29 @@ func containerCreateInternal(d *Daemon, args containerArgs) (container, error) {
+		} else if len(profileRootDiskDevices) > 1 {
+			return nil, fmt.Errorf("Container relies on profile's root disk device but conflicting devices were found")
+		}
+	}
```
This part looks a bit wrong to me.

If I read it right, the following cases are considered valid even though they shouldn't:
 1) Container has a root device called "root" without a "pool" property set
 2) Container has a root device called "root" with a "pool" property set and also inherits from profile1 with a root device called "root1" and from profile2 with a root device called "root2"

brauner replied:
The first point is probably right but the second case:
- as long as all profiles specify the same pool this isn't a problem, right? This is the premise we set out at the beginning. What should not happen is that a container is created *without*  a local root device and with conflicting root devices in its' profiles.

brauner replied:
In any case, I'm going to insert a check for that the precludes a container with multiple root disk devices in its profiles being created. We can loosen this restriction later on

stgraber commented on lxd/container_lxc.go:
```diff
@@ -210,17 +210,16 @@ func containerLXCCreate(d *Daemon, args containerArgs) (container, error) {
-	storagePool := c.StoragePool()
 	// Get the storage pool ID for the container.
-	poolID, pool, err := dbStoragePoolGet(d.db, storagePool)
+	poolID, pool, err := dbStoragePoolGet(d.db, c.storagePool)
```
Don't you need to call initStorage for that to be guaranteed to work?

stgraber replied:
Oh, I see, init() calls a new more minimal initstorage of some sort.

stgraber commented on lxd/container_lxc.go:
```diff
@@ -1369,8 +1373,26 @@ func (c *containerLXC) initLXC() error {
+	}
+	if rootDiskDevice["pool"] == "" {
+		return fmt.Errorf("No storage pool found in the containers devices.")
+	}
```
We shouldn't do that. Otherwise we run into the same problem I described before where if the container has a broken config, we can't ever fix it because all container functions call init() which calls initStoragePool() causing all of them to fail and requiring the user go fix the database by hand.

brauner replied:
We can reverse engineer the pool the container is on when it is already created but this function here needs to be called at least in `containerLXCCreate()`.

brauner replied:
- Right, what it does now, is that containerLXCCreate() parses the devices list to detect a valid storage pool. 
- init() now calls a function initStoragePool() that only looks at the database init also doesn't check for failure of initStoragePool(). It only warns when it failed to detect a valid storage pool

stgraber commented on lxd/container_lxc.go:
```diff
@@ -3224,57 +3220,83 @@ func (c *containerLXC) Update(args containerArgs, userRequested bool) error {
+			}
+			newRootDiskDeviceKey = newProfileRootDiskDevices[0]
 		}
 	}
```
Just a note that I need to review this whole chunk of code. The errors messages all look good but the new/old logic is pretty complex and I need to read through it carefully.

stgraber commented on lxd/profiles_utils.go:
```diff
@@ -0,0 +1,135 @@
+				}
+			}
+		}
+	}
```
Another note to myself to go through this chunk again in the morning. It looks like my suggested pseudo-code but my eyes are tired and I'll need to read through again to be sure that all the old/new logic is right.

stgraber commented:
jenkins: test this please

stgraber commented on lxd/container.go:
```diff
@@ -660,6 +681,37 @@ func containerCreateInternal(d *Daemon, args containerArgs) (container, error) {
+	// Check whether container has a local root device with a "pool"
+	// property set.
+	if k != "" && newLocalRootDiskDevice["pool"] == "" {
+		return nil, fmt.Errorf("A root disk device must have the \"pool\" property set.")
```
Won't that break copying from older LXD releases? Or is the code to support those copies not in this branch yet?

brauner replied:
No, it shouldn't because before containerCreateInternal() is called, createFromMigration() will set a storage pool. In my testing it worked. But I'd prefer if you did some tests yourself too. :)

stgraber replied:
@brauner ah right, that should be fine then.

stgraber replied:
@brauner where is that logic in createFromMigration()? I'm failing to see anything which sets the pool property when it's missing in that function.

brauner replied:
@stgraber, this part: https://github.com/lxc/lxd/pull/2903/files#diff-a8c4bafcd7965c655b000037f7d485adR224

```
	if localRootDiskDeviceKey == "" || localRootDiskDeviceKey != "" && v["pool"] == "" {
		// Request came without a root disk device or without the pool
		// property set. Try to retrieve this information from a
		// profile. (No need to check for conflicting storage pool
		// properties in the profiles. This will be handled by
		// containerCreateInternal().)
		for _, pName := range req.Profiles {
			_, p, err := dbProfileGet(d.db, pName)
			if err != nil {
				return InternalError(err)
			}

			k, v := containerGetRootDiskDevice(p.Devices)
			if k != "" && v["pool"] != "" {
				poolForCopyOrMove = v["pool"]
				break
			}
		}
		if poolForCopyOrMove == "" {
			pools, err := dbStoragePools(d.db)
			if err != nil {
				return InternalError(err)
			}

			if len(pools) != 1 {
				return InternalError(fmt.Errorf("No unique storage pool found."))
			}

			poolForCopyOrMove = pools[0]
		}
	}
```

stgraber commented on lxd/patches.go:
```diff
@@ -245,58 +245,146 @@ func patchStorageApi(name string, d *Daemon) error {
 		if err != nil {
-			tx.Rollback()
-			return err
+			continue
```
We probably should at least log an error here...

brauner replied:
Yep.

stgraber commented on lxd/patches.go:
```diff
@@ -846,48 +934,52 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 			if err != nil {
-				return fmt.Errorf("Failed to unmount ZFS filesystem: %s", output)
+				shared.LogWarnf("Failed to unmount ZFS filesystem via zfs unmount. Retrying with umount -l...")
+				output, err := tryExec("umount", "-l", oldContainerMntPoint)
```
syscall.Umount would probably be best here.

brauner replied:
Right, using `tryUnmount(mntPoint, syscall.MNT_DETACH)`.

stgraber commented on lxd/patches.go:
```diff
@@ -954,23 +1051,30 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 				if err != nil {
-					return fmt.Errorf("Failed to unmount ZFS filesystem: %s", output)
+					shared.LogWarnf("Failed to unmount ZFS filesystem via zfs unmount. Retrying with umount -l...")
+					output, err := tryExec("umount", "-l", oldImageMntPoint)
```
syscall.Unmount

brauner replied:
Right, using `tryUnmount(mntPoint, syscall.MNT_DETACH)`.

stgraber commented:
Left a few more comments that should be addressed.

stgraber commented:
Also, now that the patch can be re-run without wreaking havoc, we should have it run a second time to finish the migration for people who upgraded successfully with 2.9.1 and so now have broken profiles and snapshots.

Adding a second patch in the list, pointing to the same function as the first should do the trick.

stgraber commented on lxd/patches.go:
```diff
@@ -34,6 +35,7 @@ var patches = []patch{
 	{name: "leftover_profile_config", run: patchLeftoverProfileConfig},
 	{name: "network_permissions", run: patchNetworkPermissions},
 	{name: "storage_api", run: patchStorageApi},
+	{name: "storage_api_upgrade_v2", run: patchStorageApiUpgradeV2},
```
Why not simply list the same function twice?

brauner replied:
oh right.

stgraber commented on lxd/patches.go:
```diff
@@ -293,26 +295,30 @@ func patchStorageApi(name string, d *Daemon) error {
+				shared.LogWarnf("Could not commit database transaction.")
 				tx.Rollback()
-				return err
+				continue
```
Those errors should be more user friendly and should be LogErrorf as they are actual failure.

    "Failed to add a root device to profile \"%s\": %v"

Should do the trick so we can get users to grep their logs and tell us what profile failed.

brauner replied:
Hm, for what is the `%v`?

stgraber replied:
err

stgraber commented on lxd/patches.go:
```diff
@@ -245,58 +247,150 @@ func patchStorageApi(name string, d *Daemon) error {
+			pID, p, err := dbProfileGet(d.db, pName)
+			if err != nil {
+				// Do not fail.
+				continue
```
Actually, there's no reason why dbProfileGet would fail, so if that fails, it means the DB is completely broken and we should fail.

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -245,58 +247,150 @@ func patchStorageApi(name string, d *Daemon) error {
-		tx, err := dbBegin(d.db)
+	// When the either no default profile is not detected or some containers
+	// do not use the default profile, give these containers a valid local
+	// root disk device with a pool property set.
```
That comment doesn't make any sense, please fix.

brauner replied:
Hm? "When no default profile is detected or some containers do not rely on the default profile for their root disk device, these containers will be given a valid local root disk device."

stgraber replied:
Sounds good. The comment currently in the file isn't valid English, has double negative, ... :)

stgraber commented on lxd/patches.go:
```diff
@@ -245,58 +247,150 @@ func patchStorageApi(name string, d *Daemon) error {
+		c, err := containerLoadByName(d, ct)
 		if err != nil {
-			return err
+			continue
```
That's worth a LogErrorf

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -854,40 +982,44 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 		if err != nil {
+			// In case we're on a LXD instance where a a partial
+			// upgrade took place, we should still proceed.
 			shared.LogWarnf("Could not insert a storage volume for container \"%s\".", ct)
```
Should check that the error is because the entry already exists, if it is, then continue, if it's some other error, then fail.

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -907,13 +1039,13 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 		ctSnapshots, err := dbContainerGetSnapshots(d.db, ct)
 		if err != nil {
-			return err
+			continue
```
There's no reason this would fail unless the DB is broken, if the DB is broken we don't want to continue, so this should be kept as fatal.

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -924,26 +1056,31 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
 			_, err := dbStoragePoolVolumeCreate(d.db, cs, storagePoolVolumeTypeContainer, poolID, volumeConfig)
 			if err != nil {
 				shared.LogWarnf("Could not insert a storage volume for snapshot \"%s\".", cs)
-				continue
```
Same as for ZFS, we should check that the error is because of an existing DB entry, if not, fail.

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -954,23 +1091,28 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			// In case we're on a LXD instance where a a partial
+			// upgrade took place, we should still proceed.
 			shared.LogWarnf("Could not insert a storage volume for image \"%s\".", img)
-			continue
```
Same as above, should be fatal unless there's actually the right record in the DB.

stgraber commented on lxd/patches.go:
```diff
@@ -34,6 +35,7 @@ var patches = []patch{
 	{name: "leftover_profile_config", run: patchLeftoverProfileConfig},
 	{name: "network_permissions", run: patchNetworkPermissions},
 	{name: "storage_api", run: patchStorageApi},
+	{name: "storage_api", run: patchStorageApi},
```
The name must be different, the "run:" function can remain the same.

So change it to:

    {name: "storage_api_1", run: patchStorageApi},

brauner replied:
done

stgraber commented:
The comment about the error messages on failed profile update still stands. The current error messages don't tell me what profile those functions failed for, so how am I supposed to fix the problem by myself?

stgraber commented on lxd/patches.go:
```diff
@@ -596,42 +549,70 @@ func upgradeFromStorageTypeLvm(name string, d *Daemon, defaultPoolName string, d
+		if err != nil {
+			return err
+		}
 	}
```
We shouldn't just move on if the LVM is broken to the point where we can't even list the VGs...

stgraber replied:
fixed this one too? :)

brauner replied:
fixed, bailing on error now.

stgraber commented on lxd/patches.go:
```diff
@@ -854,40 +870,55 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			_, err := dbStoragePoolVolumeCreate(d.db, ct, storagePoolVolumeTypeContainer, poolID, volumeConfig)
+			if err != nil {
+				// In case we're on a LXD instance where a a partial
+				// upgrade took place, we should still proceed.
```
That comment doesn't make sense anymore.

brauner replied:
done

stgraber commented on lxd/patches.go:
```diff
@@ -854,40 +870,55 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
-		err = os.Remove(oldContainerMntPoint)
-		if err != nil {
-			return err
+		if shared.PathExists(oldContainerMntPoint) {
```
Those PathExists() checks can be dropped if you're not going to check the return value of os.Remove anyway.

brauner replied:
dropped

stgraber commented on lxd/patches.go:
```diff
@@ -921,29 +957,48 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+				_, err := dbStoragePoolVolumeCreate(d.db, cs, storagePoolVolumeTypeContainer, poolID, volumeConfig)
+				if err != nil {
+					// In case we're on a LXD instance where a a partial
+					// upgrade took place, we should still proceed.
```
Comment doesn't make sense anymore

brauner replied:
removed

stgraber commented on lxd/patches.go:
```diff
@@ -921,29 +957,48 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
-		os.RemoveAll(snapshotsPath)
+		if shared.PathExists(snapshotsPath) {
+			os.RemoveAll(snapshotsPath)
+		}
```
Same comment as above, PathExists isn't necessary if you're not going to check for the error anyway

brauner replied:
removed

stgraber commented on lxd/patches.go:
```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			_, err := dbStoragePoolVolumeCreate(d.db, img, storagePoolVolumeTypeImage, poolID, volumeConfig)
+			if err != nil {
+				// In case we're on a LXD instance where a a partial
+				// upgrade took place, we should still proceed.
```
Same as above

brauner replied:
removed

stgraber commented on lxd/patches.go:
```diff
@@ -854,40 +870,55 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
-		err = os.Remove(oldContainerMntPoint + ".zfs")
-		if err != nil {
-			return err
+		if shared.PathExists(oldContainerMntPoint + ".zfs") {
```
Here too

brauner replied:
removed

stgraber commented on lxd/patches.go:
```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			k, _ := containerGetRootDiskDevice(p.Devices)
+			if k != "" {
+				// If found, set the pool property
+				p.Devices[k]["pool"] = poolName
```
Only do that if it doesn't already have a value set. Since we will trigger this code with 2.9.2, there may be 2.9.1 users that created and configured alternate pools already.

stgraber replied:
Nevermind, I see you skip the upgrade if there are more than one pools defined. That's not perfect but I guess we can tell those users how to fix their profiles if they got that far.

stgraber replied:
Nevermind, I see you skip the upgrade if there are more than one pools defined. That's not perfect but I guess we can tell those users how to fix their profiles if they got that far.

brauner replied:
```				
if p.Devices[k]["pool"] != "" {
        continue
}
p.Devices[k]["pool"] = poolName
```

stgraber commented on lxd/patches.go:
```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+				shared.LogErrorf("Failed to commit database transaction: %s.", err)
+				tx.Rollback()
+				continue
+			}
```
Same comment as before, none of those errors tell me what profile failed to update so how am I supposed to fix things by hand?

brauner replied:
added profile name to debug output

stgraber commented on lxd/patches.go:
```diff
@@ -952,40 +1007,249 @@ func upgradeFromStorageTypeZfs(name string, d *Daemon, defaultPoolName string, d
+			// On partial upgrade the container might already have a
+			// valid root disk device entry.
+			if expandedDevices[k]["pool"] == poolName {
+				continue
```
And it may not be set to poolName, so should be skipped if set to != ""

stgraber replied:
Nevermind, I see you skip the upgrade if there are more than one pools defined. That's not perfect but I guess we can tell those users how to fix their profiles if they got that far.

brauner replied:
if localDevices[k]["pool"] != "" {
				continue
			}
			localDevices[k]["pool"] = poolName
			args.Devices = localDevices
